### DISCUSSION TITLE
Epiverse-TRACE developer space - Lesser-known reasons to prefer apply() over for loops

### DISCUSSION THREAD

**Discussion by giscus** [epiverse-trace/epiverse-trace.github.io#128] [Blog post discussion]
# Epiverse-TRACE developer space - Lesser-known reasons to prefer apply() over for loops

A place for Epiverse-TRACE developers to share their reflections, learnings, and showcase their work.

https://epiverse-trace.github.io/posts/for-vs-apply/

**Comment by cellocgw** [comment]
The index variable doesn't "Leak" anywhere.  It is defined inside its own environment, so any for-loop inside a closure will not affect any variable in the .globalenv.    but in any case, only a fool fails to set  **all** named loop indices to names never used anywhere else.

**Comment by rfhb** [comment]
> where apply functions can’t: when the next step depends on the previous one. This concept is known as recursion.

Isn't `rapply` for this purpose? See https://stat.ethz.ch/R-manual/R-devel/library/base/html/rapply.html

**Reply by Bisaloo** [reply]
The name can lead you to believe so, but no. "Recursive" in `rapply()` means it can go down in nested lists and apply the function `FUN` on internal elements.

The help page jumps straight to more complex examples but here is a simpler one:

``` r
x <- list(a = 3, b = list(c = 2))

# We can see that a and c are not at the same level of nestedness
str(x)
#> List of 2
#>  $ a: num 3
#>  $ b:List of 1
#>   ..$ c: num 2

# Yet `rapply()` manages to apply `sqrt()` on both `a` and `c`
rapply(x, sqrt)
#>        a      b.c 
#> 1.732051 1.414214
```

<sup>Created on 2023-11-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

In this example, you can see that there not interdependence between the different elements. We're really in the case of iterating a function on a complex list.

Outside of the `apply()` family, there are indeed functions to work with recursion though: `Reduce()` / `purrr::reduce()`. You can read more about this in the [dedicated section in the "Advanced R" book by Hadley Wickham](https://adv-r.hadley.nz/functionals.html#reduce), or [this blog post by Maëlle Salmon](https://masalmon.eu/2023/07/26/reduce/).

Thanks a lot for pointing out something that should have been clarified in the original post! :pray:
